File: Frontend/src/app/services/deliverer.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Order {
    id: number;
    user: any;
    deliverer?: any;
    orderItems: any[];
    totalAmount: number;
    status: string;
    paymentMethod: string;
    orderDate: string;
    deliveryAddress: string;
    deliveryPhone: string;
    deliveryName: string;
    notes?: string;
}

export interface Deliverer {
    id: number;
    username: string;
    email: string;
    nom: string;
    prenom: string;
    telephone: string;
    adresse: string;
    role: string;
}

@Injectable({
    providedIn: 'root'
})
export class DelivererService {
    private apiUrl = 'http://localhost:8082/api/deliverer';

    constructor(private http: HttpClient) { }

    /**
     * Récupère toutes les commandes confirmées (disponibles pour assignation)
     */
    getAvailableOrders(): Observable<Order[]> {
        return this.http.get<Order[]>(`${this.apiUrl}/orders/available`);
    }

    /**
     * Récupère les commandes assignées à un livreur
     */
    getDelivererOrders(delivererId: number): Observable<Order[]> {
        return this.http.get<Order[]>(`${this.apiUrl}/orders/${delivererId}`);
    }

    /**
     * Assigne une commande à un livreur
     */
    assignOrder(orderId: number, delivererId: number): Observable<Order> {
        return this.http.post<Order>(`${this.apiUrl}/assign/${orderId}/${delivererId}`, {});
    }

    /**
     * Démarre la livraison
     */
    startDelivery(orderId: number, delivererId: number): Observable<Order> {
        return this.http.post<Order>(`${this.apiUrl}/start/${orderId}/${delivererId}`, {});
    }

    /**
     * Confirme la livraison (check-in)
     */
    confirmDelivery(orderId: number, delivererId: number): Observable<Order> {
        return this.http.post<Order>(`${this.apiUrl}/confirm/${orderId}/${delivererId}`, {});
    }

    /**
     * Récupère tous les livreurs
     */
    getAllDeliverers(): Observable<Deliverer[]> {
        return this.http.get<Deliverer[]>(`${this.apiUrl}/all`);
    }
}

File: Backend/src/main/java/com/iset/Backend/E_commerce/Services/DelivererService.java
package com.iset.Backend.E_commerce.Services;

import com.iset.Backend.E_commerce.Entities.Order;
import com.iset.Backend.E_commerce.Entities.User;
import com.iset.Backend.E_commerce.Repositories.OrderRepository;
import com.iset.Backend.E_commerce.Repositories.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class DelivererService {

    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private UserRepository userRepository;

    /**
     * Récupère toutes les commandes confirmées (prêtes à être assignées)
     */
    public List<Order> getConfirmedOrders() {
        return orderRepository.findByStatus(Order.OrderStatus.CONFIRMED);
    }

    /**
     * Récupère les commandes assignées à un livreur spécifique
     */
    public List<Order> getDelivererOrders(Long delivererId) {
        return orderRepository.findByDelivererId(delivererId);
    }

    /**
     * Assigne une commande à un livreur
     */
    @Transactional
    public Order assignOrder(Long orderId, Long delivererId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Commande non trouvée"));

        User deliverer = userRepository.findById(delivererId)
                .orElseThrow(() -> new RuntimeException("Livreur non trouvé"));

        if (deliverer.getRole() != User.Role.DELIVERER) {
            throw new RuntimeException("L'utilisateur n'est pas un livreur");
        }

        if (order.getStatus() != Order.OrderStatus.CONFIRMED) {
            throw new RuntimeException("La commande doit être confirmée pour être assignée");
        }

        order.setDeliverer(deliverer);
        order.setStatus(Order.OrderStatus.ASSIGNED);

        return orderRepository.save(order);
    }

    /**
     * Démarre la livraison (le livreur a pris en charge la commande)
     */
    @Transactional
    public Order startDelivery(Long orderId, Long delivererId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Commande non trouvée"));

        if (order.getDeliverer() == null || order.getDeliverer().getId() != delivererId) {
            throw new RuntimeException("Cette commande n'est pas assignée à ce livreur");
        }

        if (order.getStatus() != Order.OrderStatus.ASSIGNED) {
            throw new RuntimeException("La commande doit être assignée pour démarrer la livraison");
        }

        order.setStatus(Order.OrderStatus.DELIVERING);

        return orderRepository.save(order);
    }

    /**
     * Confirme la livraison (check-in)
     */
    @Transactional
    public Order confirmDelivery(Long orderId, Long delivererId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Commande non trouvée"));

        if (order.getDeliverer() == null || order.getDeliverer().getId() != delivererId) {
            throw new RuntimeException("Cette commande n'est pas assignée à ce livreur");
        }

        if (order.getStatus() != Order.OrderStatus.DELIVERING) {
            throw new RuntimeException("La commande doit être en cours de livraison");
        }

        order.setStatus(Order.OrderStatus.DELIVERED);

        return orderRepository.save(order);
    }

    /**
     * Récupère tous les livreurs
     */
    public List<User> getAllDeliverers() {
        return userRepository.findByRole(User.Role.DELIVERER);
    }
}
